### DISCUSSION TITLE
uint64_t and int64

### DISCUSSION THREAD

**Discussion by kiminh** [Qihoo360/tensornet#63] [General] [answered]
as the author said: "convert all features into uint64 first, you can choose hash function like murmurhash do that."

for example, I use murmurhash to hash "1:23"(fea_val is 1, fea_slot is 23) to uint64 18024615837690486380, while for int64, it is -422128236019065236.

but my question is that in tfrecord only support int64 (tf.train.Int64List). then how can I put the uint64 hash value into tfrecord.
in your example, the feasign range is (1,2^63) not (1,2^64)

https://github.com/Qihoo360/tensornet/blob/master/examples/gen_example_data.py#L39. 
```
def generate_data(dt, name):
    slots = [("1", 1), ("2", 1), ("3", 1), ("4", 3)]
    count = 12000

    sign_pool = {}

    for slot, slot_len in slots:
        sign_pool[slot] = [random.randint(1, 2**63) for i in range(10)]

    # test slot 1 as dense feature
    sign_pool["1"] = [1, 2, 3]

    with tf.io.TFRecordWriter("./data/{}/tf-part.{}".format(dt, name)) as writer:
        for i in range(count):
            example = serialize_example(slots, sign_pool)
            writer.write(example)
```

```
import com.google.common.hash.Hashing;
import org.junit.Test;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;

public class StringToHash {

    public static String align(byte[] bytes) {
        return Long.toUnsignedString(padToLong(bytes));
    }

    public static long padToLong(byte[] bytes) {
        long retVal = (bytes[0] & 0xFF);
        for (int i = 1; i < Math.min(bytes.length, 8); i++) {
            retVal |= (bytes[i] & 0xFFL) << (i * 8);
        }
        return retVal;
    }
    private String orNull(Object v) {
        if (v != null) {
            return v.toString();
        }
        return null;
    }

    private String encode(String fv, String seed) {
        return fv + ":" + seed;
    }

    public byte[] doHash(String v) {
        return Hashing.murmur3_128().hashString(v, StandardCharsets.UTF_8).asBytes();
    }

    public Long doHash_1(String v) {
        return Hashing.murmur3_128().hashString(v, StandardCharsets.UTF_8).asLong();
    }


    @Test
    public void test() {
        String fea_val = "1";
        String slot = "23";
        String align = align(doHash(encode(orNull(fea_val), slot)));
        System.out.println(align);
        // 18024615837690486380
        Long hash = doHash_1(encode(orNull(fea_val), slot));
        System.out.println(hash);
        // -422128236019065236
     }
}   
```

**Comment by gzm55** [comment, marked as answer]
uint64 and in64 are identical in memory and cpu, i.e., when use c++, we can assert:
```
assert((int64_t)18024615837690486380 == -422128236019065236);
assert((uint64_t)-422128236019065236 == 18024615837690486380);
```

So, if not converting  to String type, the two types are interchangeable.

**Reply by kiminh** [reply]
> uint64 and in64 are identical in memory and cpu, i.e., when use c++, we can assert:
> 
> ```
> assert((int64_t)18024615837690486380 == -422128236019065236);
> assert((uint64_t)-422128236019065236 == 18024615837690486380);
> ```
> 
> So, if not converting to String type, the two types are interchangeable.

so in my example, I can put -422128236019065236 into TFRecord，is that right

**Reply by gzm55** [reply]
yes

**Reply by kiminh** [reply]
> yes

appreciate

**Comment by kiminh** [comment]
alternatively，maybe I can reduce the feasign size from 2^64 to 2^{63} or 2^{60} （whatever size smaller than 2^{63}) using hash + mod or mask method to make sure the positive feasign which put into tfrecord @gzm55 

1. tensorflow hash bucket method, which is hash + mod
```
 max_b = (1 << 63) - 1
'mov': tf.strings.to_hash_bucket_fast([x['movie_title']], max_b)
```
2. mask
```
uint64 Uint60Mask= 0x0fffffffffffffff #(2^60-1)
uint64 murmurhash_value = 18024615837690486380
murmurhash_value && Uint60Mask
```
